### PR TITLE
Update AST - new import syntax etc.

### PR DESCRIPTION
**Pull request by BinarySoftware**
### Pull Request Description
Updates AST and all necessary stuff for newest parser from Engine, making our lives much easier, without the need to use backported branch.

### Checklist
Please include the following checklist in your PR:

- [x] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [x] All code has automatic tests where possible.
- [x] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] ~All code has been manually tested in the "debug/interface" scene.~
- [x] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [x] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,7 +16,8 @@ read the notes of the `Enso 2.0.0-alpha.1` release.
   now on all supported platforms.
 - [You can now see data frames in the table visualisation][1181]. Big tables get truncated to 2000 
   entries.
-  
+- [Updated AST][1186]. Introducing new import/export syntax, as well as annotations.
+
```
Comment (farmaazon) [inline CHANGELOG.md]: I think this should not be written in changelog. It means nothing to the users.

File: src/rust/ide/lib/parser/build.rs [hunk 2]
```diff
@@ -27,7 +27,7 @@ const PARSER_PATH: &str = "./pkg/scala-parser.js";
 /// Commit from `enso` repository that will be used to obtain parser from.
 /// FIXME: Using commit not from Engine's main branch but from a backported branch as the AST in
 ///        this project hasn't been updated to conform to latest scala parser (#732).
-const PARSER_COMMIT: &str = "9d848fc59bfbab8aa202da71dc899759817dff0f";
+const PARSER_COMMIT: &str = "5e309bddcbec33cfbd150fcb8a16b45192cf5189";
```
Comment (farmaazon) [inline src/rust/ide/lib/parser/build.rs]: Remove the FIXME part of docs, as we don't use the backported branch AFAIK.

File: src/rust/ide/lib/parser/tests/parsing.rs [hunk 3]
```diff
@@ -457,3 +456,19 @@ fn block_roundtrip() {
...
+    use ast::*;
+    if let Shape::Infix(Infix {larg:_larg,loff:_loff,opr,roff:_roff,rarg}, ..) = ast.shape() {
+        if let Shape::Opr(Opr{..}) = opr.shape() {
+            if let Shape::Var(Var{..}) = rarg.shape() {
+                return;
+            }
+        }
+    }
+    panic!("Should have matched into return.");
+}
```
Comment (farmaazon) [inline src/rust/ide/lib/parser/tests/parsing.rs]: add a newline

**Review by farmaazon — APPROVED** [review]
Apply comments before merge.

File: CHANGELOG.md (hunk 1 above)
Comment (BinarySoftware) [reply CHANGELOG.md]: Well, something needs to be written, otherwise CI check will fail. Any suggestions?

Comment (farmaazon) [reply CHANGELOG.md]: Yes. Just add `[ci no changelog needed]` in the commit message (see https://github.com/enso-org/ide/actions/runs/555353080)

File: CHANGELOG.md [hunk 4]
```diff
@@ -46,6 +46,7 @@ read the notes of the `Enso 2.0.0-alpha.1` release.
 <br/>
 
 
+
```
Comment (mwu-tow) [inline CHANGELOG.md]: spurious newline?

File: src/rust/ide/lib/ast/impl/src/crumbs.rs [hunk 5]
```diff
@@ -246,14 +246,9 @@ pub enum BlockCrumb {
...
-pub struct ImportCrumb(pub Box<Crumb>);
-
-impl ImportCrumb {
-    #[allow(missing_docs)]
-    pub fn new(crumb:impl Into<Crumb>) -> Self {
-        Self(Box::new(crumb.into()))
-    }
+#[derive(Clone,Copy,Debug,PartialEq,Eq,PartialOrd,Ord,Hash)]
+pub enum ImportCrumb {
+    Path {index:usize}
```
Comment (mwu-tow) [inline src/rust/ide/lib/ast/impl/src/crumbs.rs]: This should also cover other `T`-dependent fields introduced to `Import` shape.
Other new shapes should also get their crumb classes, most importantly `Annotate`.
Basically the intent was to keep all the shapes and crumbs structure compatible.

File: src/rust/ide/lib/parser/tests/macros.rs [hunk 6]
```diff
@@ -31,6 +31,8 @@ fn import_utilities() {
     expect_import("import Foo");
     expect_import("import Foo.Bar");
     expect_import("import Foo.Bar.Baz");
+    expect_import("from Foo import Bar");
+    expect_import("from Foo import all hiding Bar");
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/macros.rs:35]: We should also try going for the trickiest (as in using all possible fields in a non-empty way) imports possible here.

File: src/rust/ide/lib/parser/tests/parsing.rs [hunk 7]
```diff
@@ -457,3 +456,19 @@ fn block_roundtrip() {
         roundtrip_program(program);
     }
 }
+
+/// Fixed test case, where we used to get invalid suffix, instead of infix.
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/parsing.rs]: There's no need to document test as "fixed". I just document "unfixed" ones with FIXMEs.

File: src/rust/ide/lib/parser/tests/parsing.rs [hunk 8]
```diff
@@ -457,3 +456,19 @@ fn block_roundtrip() {
...
+#[wasm_bindgen_test]
+fn extension_operator_methods() {
+    let ast = parser::Parser::new_or_panic().parse_line("Int.+").unwrap();
+
+    use ast::*;
+    if let Shape::Infix(Infix {larg:_larg,loff:_loff,opr,roff:_roff,rarg}, ..) = ast.shape() {
+        if let Shape::Opr(Opr{..}) = opr.shape() {
+            if let Shape::Var(Var{..}) = rarg.shape() {
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/parsing.rs]: This does not look correct, I believe the shape in within `rarg` should be an `Opr`.
`+` is not a `Var` identifier.

File: src/rust/ide/lib/parser/tests/parsing.rs [hunk 9]
```diff
@@ -457,3 +456,19 @@ fn block_roundtrip() {
...
     }
 }
+
+/// Fixed test case, where we used to get invalid suffix, instead of infix.
+#[wasm_bindgen_test]
+fn extension_operator_methods() {
+    let ast = parser::Parser::new_or_panic().parse_line("Int.+").unwrap();
+
+    use ast::*;
+    if let Shape::Infix(Infix {larg:_larg,loff:_loff,opr,roff:_roff,rarg}, ..) = ast.shape() {
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/parsing.rs]: All the matched fields (here and below) should be asserted to contain sensible values.

File: src/rust/ide/lib/parser/tests/macros.rs (hunk 6 above)
Comment (mwu-tow) [reply src/rust/ide/lib/parser/tests/macros.rs:35]: We could just take examples from the docs.

File: src/rust/ide/lib/parser/tests/macros.rs [hunk 10]
```diff
@@ -31,6 +31,8 @@ fn import_utilities() {
...
 
     expect_not_import("type Foo");
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/macros.rs:38]: We should test that exports are not recognized as imports.

File: src/rust/ide/lib/ast/impl/src/macros.rs [hunk 11]
```diff
@@ -30,7 +32,7 @@ pub fn ast_as_import_match(ast:&Ast) -> Option<known::Match> {
 pub fn is_match_import(ast:&known::Match) -> bool {
     let segment = &ast.segs.head;
     let keyword = crate::identifier::name(&segment.head);
-    keyword.contains_if(|str| *str == IMPORT_KEYWORD)
+    keyword.contains_if(|str| *str == QUALIFIED_IMPORT_KEYWORD || *str == UNQUALIFIED_IMPORT_KEYWORD)
```
Comment (mwu-tow) [inline src/rust/ide/lib/ast/impl/src/macros.rs]: Won't that recognize the unqualified export as an unqualified import?
Consider: `from Base.Data.List export Cons, Nil, from_vector`

File: src/rust/ide/lib/parser/tests/parsing.rs [hunk 12]
```diff
@@ -361,7 +361,7 @@ impl Fixture {
 
     /// Tests parsing a number of sample macro usages.
     ///
-    /// As macros geneerate usually really huge ASTs, this test only checks
+    /// As macros generate usually really huge ASTs, this test only checks
     /// that we are able to deserialize the response and that it is a macro
     /// match node. Node contents is not covered.
     fn deserialize_macro_matches(&mut self) {
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/parsing.rs:374]: Like we have these, we should also add a method that tests deserialization of the newly introduced `Annotation` shape. If it is not possible/feasible for some reason (tbh I have no idea what `Annotation` shape is), then it should be mentioned in the comment in the `run` function, along with other not separately covered types.

Comment (mwu-tow) [reply src/rust/ide/lib/parser/tests/parsing.rs:374]: Also, `deserialize_macro_matches` should be expanded with examples of code using the newly supported features (import/exports with renames, and this Documented thingy)

**Review by mwu-tow — CHANGES_REQUESTED** [review]

File: src/rust/ide/lib/parser/tests/parsing.rs (hunk 12 above)
Comment (BinarySoftware) [reply src/rust/ide/lib/parser/tests/parsing.rs:374]: From call : parser will never ever return AST.Documented, it should be in AST as the engine and IDE's AST should always be compliant, but as we'll never ever get to documented, there's no need for tests (as we cant test it even) or for crumbs.

File: src/rust/ide/lib/parser/tests/parsing.rs (hunk 8 above)
Comment (BinarySoftware) [reply src/rust/ide/lib/parser/tests/parsing.rs]: This will go back to bugs

File: src/rust/ide/lib/parser/tests/parsing.rs [hunk 13]
```diff
@@ -359,9 +359,16 @@ impl Fixture {
         });
     }
 
+    fn deserialize_annotation(&mut self) {
+        self.test_shape("@Tail_call",|annotation: &Annotation| {
+            let expected_var = Annotation {name:"@Tail_call".into()};
```
Comment (mwu-tow) [inline src/rust/ide/lib/parser/tests/parsing.rs]: `expected_var` should be named rather like `expected_annotation`

**Review by mwu-tow — APPROVED** [review]

**Pull request merged**